velox/exec/benchmarks/HashJoinListResultBenchmark.cpp

```diff
@@ -0,0 +1,465 @@
+#include "velox/functions/prestosql/aggregates/RegisterAggregateFunctions.h"
+#include "velox/vector/tests/utils/VectorTestBase.h"
+
+#include <folly/Benchmark.h>
+#include <folly/executors/CPUThreadPoolExecutor.h>
+#include <folly/init/Init.h>
+#include <gtest/gtest.h>
+#include <memory>
+
+using namespace facebook::velox;
+using namespace facebook::velox::exec;
+using namespace facebook::velox::test;
+
+namespace {
+struct HashTableBenchmarkParams {
+  HashTableBenchmarkParams() = default;
+
+  HashTableBenchmarkParams(
+      BaseHashTable::HashMode mode,
+      TypePtr buildType,
+      int64_t hashTableSize,
+      int64_t probeSize,
+      const std::vector<std::pair<int32_t, int32_t>>&
+          keyRepeatTimesDistribution)
+      : mode(mode),
+        buildType(buildType),
+        hashTableSize(hashTableSize),
+        probeSize(probeSize),
+        keyRepeatTimesDistribution(std::move(keyRepeatTimesDistribution)) {
+    int32_t distSum = 0;
+    buildSize = 0;
+    buildKeyRepeat.reserve(keyRepeatTimesDistribution.size());
+    for (auto dist : keyRepeatTimesDistribution) {
+      if (dist.first > 100 || dist.first < 0 || dist.second < 0) {
+        VELOX_FAIL(
+            "Bad distribution: [distribution:{}, key duplicate count:{}]",
+            dist.first,
+            dist.second);
+      }
+      buildSize += (hashTableSize * dist.first / 100) * (dist.second + 1);
+      distSum += dist.first;
+      buildKeyRepeat.emplace_back(
+          std::make_pair(hashTableSize * distSum / 100, dist.second));
+    }
+    VELOX_CHECK_EQ(distSum, 100, "Sum of distributions should be 100");
+
+    if (hashTableSize > BaseHashTable::kArrayHashMaxSize &&
+        mode == BaseHashTable::HashMode::kArray) {
+      VELOX_FAIL("Bad hash mode.");
+    }
+
+    numDependentFields = buildType->size() - 1;
+    if (mode == BaseHashTable::HashMode::kNormalizedKey) {
+      extraValue = (2L << 20) + 100;
+    } else if (mode == BaseHashTable::HashMode::kHash) {
+      extraValue = std::numeric_limits<int64_t>::max() - 1;
+    } else {
+      extraValue = 0;
+    }
+    std::string modeString = mode == BaseHashTable::HashMode::kArray ? "array"
+        : mode == BaseHashTable::HashMode::kHash                     ? "hash"
+                                                 : "normalized key";
+    std::stringstream distStr;
+    for (auto dist : keyRepeatTimesDistribution) {
+      distStr << fmt::format("{}%:{};", dist.first, dist.second);
+    }
+    title = fmt::format(
+        "{}_{}_{}", modeString, numDependentFields + 1, distStr.str());
+  }
+
+  // Title for reporting
+  std::string title;
+
+  // Expected mode.
+  BaseHashTable::HashMode mode;
+
+  // Type of build & probe row.
+  TypePtr buildType{ROW({"k1"}, {BIGINT()})};
+
+  // Distinct rows in the table.
+  int64_t hashTableSize;
+
+  // Number of probe rows.
+  int64_t probeSize;
+
+  // Number of build rows.
+  int64_t buildSize;
+
+  // This is used to control the hash mode.
```

Comment:
Would you clarify a bit more? It is not obvious what it means to control the hash mode and how that is achieved with a single integer.

Reply:
Added coments for this property. This property will be added to the hash table. We set different values based on the expected hash mode.
array: 0
normalized key,: 'kArrayHashMaxSize' + 100
hash: maximum value of int64_t -1